text/0000-template.md

```diff
@@ -0,0 +1,78 @@
+- Start Date: 2019-02-13
+- Relevant Teams: Ember.js, Learning
+- RFC PR: https://github.com/emberjs/rfcs/pull/445
+- Tracking: (leave this empty)
+
+# Deprecate `{{with}}`
+
+## Summary
+
+The `{{with}}` helper has always had slightly confusing conditional semantics, this was one of the motivators for [introducing](https://github.com/emberjs/rfcs/blob/master/text/0286-block-let-template-helper.md) the easier to understand `{{let}}` helper. Now that `{{let}}` exists, the remaining use case for using `{{with}}` is for its unique conditional semantics. These conditional semantics can be cleanly represented with a combination of `{{let}}` and `{{if}}` statements so ee should deprecate `{{with}}`.
+
+## Motivation
+
+The difference between `{{let}}` and `{{with}}` is with how they handle conditional arguments. The `{{let}}` helper's block content is always rendered, regardless of its parameters. In contrast, `{{with}}` only renders its main block when the first position parameter is truthy. For example:
+
+```hbs
+{{#with "Alex" as |value|}}
+  {{value}} is truthy
+{{else}}
+  The first positional param was falsy
+{{/with}}
+```
+
+Will render "[Alex] is truthy".
+
+```hbs
+{{#with false as |value|}}
+  {{value}} is truthy
+{{else}}
+  The first positional param was falsy
+{{/with}}
+```
+
+Will render "The first positional param was falsy".
+
+The conditional arguments behavior of `{{with}}` can easily be replicated using a combination of `{{let}}` and `{{if}}` in a way that's easily readable:
+
+```hbs
+{{#let "Alex" as |value|}}
+  {{#if value}}
+    {{value}} is truthy
+  {{else}}
+    The first positional param was falsy
+  {{/if}}
+{{/let}}
+```
+
+## Detailed design
+
+We'll create an AST transform in [`packages/ember-template-compiler`](https://github.com/emberjs/ember.js/tree/master/packages/ember-template-compiler) which will emit a deprecation warning for all uses of `{{with}}`. The deprecation warning will be something like:
+
+```
+Using `{{with}}` is deprecated, please use a combination of `{{let}}` and `{{if}}` instead.
```

Comment:
I think that the message should only suggest `if` _if_ there is an `{{else`, what do you think?

Reply:
Perhaps some people rely on the main `{{with}}` block not rendering when the first argument in falsy? If so, I believe that they would need to migrate to using `{{let}}` and `{{if}}`.

Reply:
Ya, I totally agree. My _guess_ (and anecdotally what I've found when teaching/pairing) is that most folks are **surprised** by the current behavior and actually do want "render this block all the time".

I think I'd propose the following "choose your own adventure" for the deprecation guide on migrating away from `{{with` usage:

* if you always want to render a given block unconditionally: use `{{let`
* if you want to render a given block conditionally: use `{{let` + `{{if`
* if you want to render a given block conditionally, otherwise rendering an alternate block: use `{{let` + `{{if` + `{{else`

Reply:
👍 sounds good, thanks

Reply:
@rwjblue I've added the deprecation guide content to this RFC